xdsl/irdl/attributes.py

```diff
@@ -371,11 +371,29 @@ def decorator(cls: TypeAttributeInvT) -> TypeAttributeInvT:
     return decorator(cls)
 
 
+# Cannot subclass ABC as it conflicts with Enum's metaclass.
+class ConstraintConvertible(Generic[AttributeCovT]):
+    """
+    Abstract superclass for values that have corresponding Attribute constraints.
+    """
+
+    @staticmethod
+    @abstractmethod
+    def base_constr() -> AttrConstraint[AttributeCovT]:
+        """The constraint for this class."""
+
+    @abstractmethod
+    def constr(self) -> AttrConstraint[AttributeCovT]:
+        """The constraint for this instance."""
+
+
 IRDLAttrConstraint: TypeAlias = (
     AttrConstraint[AttributeInvT]
     | AttributeInvT
     | type[AttributeInvT]
     | "TypeForm[AttributeInvT]"
+    | type[ConstraintConvertible[AttributeInvT]]
+    | ConstraintConvertible[AttributeInvT]
```

Comment:
So is the idea that you can just pass in the enum directly instead of the Attribute that wraps the enum?

Reply:
Yep exactly

Reply:
Would you be able to point at where this is used in the big PR (or push a follow up PR). It's not clear to me how this all fits together yet. Shouldn't `Attribute` itself be `ConstraintConvertible`?

Reply:
Will `int` also be `ConstraintConvertible`?

Reply:
No because we can't retroactively conform it to that, it'll be special cased, and the assumption is that you always want to use IntAttr for int, but you don't know which attribute to use for a given enum

Reply:
Hmm, I'm not sure I really understand this then, is this abstract class just for enums?